---
title: "DMG_Analysis"
author: "Javier Rodriguez Casariego"
date: "10/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(plotrix) 
library(ggplot2)
library(gridExtra)
library(pheatmap)
library(dplyr)
library(tidyverse)
library(tidyr)
library(topGO)
library(goseq) #BiocManager:: install('goseq')
library(genefilter)
library(gplots)
library(cowplot)
library(lsmeans) #install.packages('lsmeans')
library(ontologyIndex) #BiocManager:: install('ontologyIndex')
library(ontologySimilarity) #BiocManager:: install('ontologySimilarity')
library(data.table)
library(RColorBrewer)
library(colorRamps)
library(GSEABase)
```


Methylation Analysis
## Loading genomic and annotation
```{r}
#load sample information
sample.info <- read.csv("Treatment_metadata.csv", header=T, sep=",", na.string="NA", stringsAsFactors = F) #read in info
colnames(sample.info)[1] <- "Sample.ID"
samp <- sample.info$Sample.ID # set sample info
#load genes gff
Genes <- read.csv("Genome/Mcav.GFFannotation.gene.gff", header=F, sep="\t", na.string="NA", skip=3, stringsAsFactors = F) #read in data file
Genes <- Genes[,c(1,4,5,9)] #select desired columns only
colnames(Genes) <- c("scaffold", "start", "stop", "gene") #rename columns
Genes$gene <- gsub(";.*","",Genes$gene) #remove extra characters
Genes$gene <- gsub("ID=","",Genes$gene) #remove extra characters
#Load annotation file
Annot <- read.csv("Genome/Mcavernosa_trinotate_annotation_report.xls.txt", header=TRUE, sep="\t", na.string="NA", stringsAsFactors = F, skip=0)
colnames(Annot)[1] <- "gene"
Annot <- merge(Annot, Genes, by="gene")
GO.data <- Annot[,c(1,13:14)] #select names and GOs

GO.data_blast <- Annot[,c(1,13)] #select names and GOs
GO.data_pfam <- Annot[,c(1,14)] #select names and GOs

splitted_blast <- strsplit(as.character(GO.data_blast$gene_ontology_blast), "`") #split into multiple GO ids
Gene.GO_blast.IDs <- data.frame(v1 = rep.int(GO.data_blast$gene, sapply(splitted_blast, length)), v2 = unlist(splitted_blast)) #list all genes with each of their GO terms in a single row 
colnames(Gene.GO_blast.IDs) <- c("gene", "GO.IDs") #rename columns
Gene.GO_blast.IDs$GO.IDs[which(Gene.GO_blast.IDs$GO.IDs == ".")] <- NA #replace NA with unknown
Gene.GO_blast.IDs$GO.IDs <- as.character(Gene.GO_blast.IDs$GO.IDs)
Gene.GO_blast.IDs$GO.IDs <- substr(Gene.GO_blast.IDs$GO.IDs, 1, 10)

splitted_pfam <- strsplit(as.character(GO.data_pfam$gene_ontology_pfam), "`") #split into multiple GO ids
Gene.GO_pfam.IDs <- data.frame(v1 = rep.int(GO.data_pfam$gene, sapply(splitted_pfam, length)), v2 = unlist(splitted_pfam)) #list all genes with each of their GO terms in a single row 
colnames(Gene.GO_pfam.IDs) <- c("gene", "GO.IDs") #rename columns
Gene.GO_pfam.IDs$GO.IDs[which(Gene.GO_pfam.IDs$GO.IDs == ".")] <- NA #replace NA with unknown
Gene.GO_pfam.IDs$GO.IDs <- as.character(Gene.GO_pfam.IDs$GO.IDs)
Gene.GO_pfam.IDs$GO.IDs <- substr(Gene.GO_pfam.IDs$GO.IDs, 1, 10)

Gene.GO.IDs <-rbind(Gene.GO_blast.IDs, Gene.GO_pfam.IDs)
Gene.GO.IDs <- Gene.GO.IDs[order(Gene.GO.IDs$gene),]

Gene.GO.IDs$seq_value <- ave(Gene.GO.IDs$gene, Gene.GO.IDs$gene, FUN=seq_along)

Gene.GO.IDs.wide <- Gene.GO.IDs %>%
  pivot_wider(names_from = seq_value, values_from = GO.IDs)

Gene.GO.IDs.clean <- Gene.GO.IDs.wide %>%
  unite(GO.IDs, 2:218, sep = ";", remove = TRUE, na.rm = T)

gene.IDs <- unique(Gene.GO.IDs$gene)
GO.IDs <-unique(Gene.GO.IDs$GO.IDs)
```

Generate slim terms for GOs
```{r}
library(foreach)
library(doParallel)

fl <- system.file("extdata", "goslim_generic.obo", package="GSEABase")
slim <- getOBOCollection(fl)

#setup parallel backend to use many processors
cores=detectCores()
cl <- makeCluster(cores[1]-1) #not to overload your computer
registerDoParallel(cl)

gene_slims <- foreach(i=1:length(Gene.GO.IDs.clean$gene), .combine=rbind, .packages = c("GSEABase", "topGO")) %dopar% {
    print(i)
  if(!is.na(Gene.GO.IDs.clean$GO.IDs[i])){
    GO_IDs <- unlist(strsplit(Gene.GO.IDs.clean$GO.IDs[i], ";"))
    GO_collection <- GOCollection(GO_IDs)
    slims_bp <- data.frame(goSlim(GO_collection, slim, "BP"))
    slims_bp$gene <- Gene.GO.IDs.clean$gene[i]
    slims_bp <- slims_bp[which(slims_bp$Count!=0),]
  } 
}
#stop cluster
stopCluster(cl)
    
gene_slims$Term <- gsub("nucleobase-containing compound cata...","nucleobase-containing compound catabolic process", gene_slims$Term)
  
gene_slims$Term <- gsub("generation of precursor metabolites...","generation of precursor metabolites and energy", gene_slims$Term)
    
gene_slims$Term <- gsub("cytoskeleton-dependent intracellula...","cytoskeleton-dependent intracellular transport", gene_slims$Term)
    
gene_slims$Term <- gsub("cellular geneein modification proce...","cellular geneein modification process", gene_slims$Term)
    
gene_slims$Term <- gsub("cellular nitrogen compound metaboli...","cellular nitrogen compound metabolic process", gene_slims$Term)
    
gene_slims$Term <- gsub("cellular nitrogen compound metaboli...","cellular nitrogen compound metabolic p", gene_slims$Term)
    
gene_slims$Term <- gsub("cellular amino acid metabolic proce...","cellular amino acid metabolic process", gene_slims$Term)
    
gene_slims$Term <- gsub("anatomical structure formation invo...","anatomical structure formation involved in morphogenesis", gene_slims$Term)

gene_slims$Term <- gsub("protein modification proce...","protein modification process", gene_slims$Term)

gene_slims$Term <- gsub("cell wall organization or biogenesi...","cell wall organization or biogenesis", gene_slims$Term)


#reformat data frame from long to wide with slim terms separated by commas
gene_slims_summry_bp <- gene_slims %>% dplyr::group_by(gene) %>% dplyr::summarise(GO_slim_BP = toString(Term)) 
#convert commas to semi colons
gene_slims_summry_bp$GO_slim_BP <- gsub(",",";",gene_slims_summry_bp$GO_slim_BP)

#do the same for CC
cores=detectCores()
cl <- makeCluster(cores[1]-1) #not to overload your computer
registerDoParallel(cl)


gene_slims <- foreach(i=1:length(Gene.GO.IDs.clean$gene), .combine=rbind, .packages = c("GSEABase", "topGO")) %dopar% {
    print(i)
   if(!is.na(Gene.GO.IDs.clean$GO.IDs[i])){
    GO_IDs <- unlist(strsplit(Gene.GO.IDs.clean$GO.IDs[i], ";"))
    GO_collection <- GOCollection(GO_IDs)
    slims_cc <- data.frame(goSlim(GO_collection, slim, "CC"))
    slims_cc$gene <- Gene.GO.IDs.clean$gene[i]
    slims_cc <- slims_cc[which(slims_cc$Count!=0),]
  }
}
stopCluster(cl)

#reformat data frame from long to wide with slim terms separated by commas
gene_slims_summry_cc <- gene_slims %>% dplyr::group_by(gene) %>% dplyr::summarise(GO_slim_CC = toString(Term)) 
#convert commas to semi colons
gene_slims_summry_cc$GO_slim_CC <- gsub(",",";",gene_slims_summry_cc$GO_slim_CC)

#do the same for MF
cores=detectCores()
cl <- makeCluster(cores[1]-1) #not to overload your computer
registerDoParallel(cl)

gene_slim <- foreach(i=1:length(Gene.GO.IDs.clean$gene), .combine=rbind, .packages = c("GSEABase", "topGO")) %dopar% {
    print(i)
  if(!is.na(Gene.GO.IDs.clean$GO.IDs[i])){
    GO_IDs <- unlist(strsplit(Gene.GO.IDs.clean$GO.IDs[i], ";"))
    GO_collection <- GOCollection(GO_IDs)
    slims_mf <- data.frame(goSlim(GO_collection, slim, "MF"))
    slims_mf$gene <- Gene.GO.IDs.clean$gene[i]
    slims_mf <- slims_mf[which(slims_mf$Count!=0),]
  }
}
stopCluster(cl)
#reformat data frame from long to wide with slim terms separated by commas
gene_slims_summry_mf <- gene_slims %>% dplyr::group_by(gene) %>% dplyr::summarise(GO_slim_MF = toString(Term)) 
#convert commas to semi colons
gene_slims_summry_mf$GO_slim_MF <- gsub(",",";",gene_slims_summry_mf$GO_slim_MF)
#merge GO slim tables from different categories together
gene_slims_summry <- merge(gene_slims_summry_bp, gene_slims_summry_cc, by = "gene", all = T)
gene_slims_summry <- merge(gene_slims_summry, gene_slims_summry_mf, by = "gene", all = T)
```

merge back with gene annotation table
```{r}
all_GO_slim <- merge(Genes,gene_slims_summry, by = "gene" , all = T)
all_GO_slim <- merge(all_GO_slim, Gene.GO.IDs.clean, by = "gene", all = T)
```


write out file
```{r}
write.table(all_GO_slim, "Genome/Mcav_annotation_gene_slims.tab", sep = "\t", quote = F, row.names = F)
```


```{r}
GoSlims <- read.csv("Genome/GO-GOslim.sorted", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = F, skip=0)
colnames(GoSlims) <- c("GO.IDs", "GO.Term", "GO.Slim.Term", "Cat") #rename columns
Gene.GO.IDs.slims <- merge(Gene.GO.IDs, GoSlims, by="GO.IDs", all = TRUE)
Gene.GO.IDs.slims <- Gene.GO.IDs.slims[-1,]

```


##############################################Load filtered methylation counts##################################
```{r}

meth.data <- list.files(path = "OSF/1.2alignment/", pattern = ".bed$", full.names=TRUE) %>%
  set_names(.) %>% 
  map_dfr(read.csv,.id="Sample.ID", header=FALSE, sep="\t", na.string="NA", stringsAsFactors = FALSE) %>% 
  dplyr::select(-c(V3,V7:V14)) %>%
  group_by(Sample.ID)
colnames(meth.data) <- c("Sample.ID", "scaffold", "position","per.meth","meth","unmeth","gene")
meth.data$gene <- gsub(";.*","",meth.data$gene) #remove extra characters
meth.data$gene <- gsub("ID=","",meth.data$gene) #remove extra characters
meth.data$Sample.ID <- gsub("OSF/1.2alignment/","",meth.data$Sample.ID) #remove extra characters
meth.data$Sample.ID <- gsub("_.*","",meth.data$Sample.ID) #remove extra characters 
MD.All <- merge(meth.data, sample.info, by="Sample.ID")
MD.All <- MD.All %>%
  mutate(sym = recode(trt1, b = "D", c = "C"),
         treatment = paste(sym, trt2, sep = ""))

```


#GO identification for background set
```{r}
#generate list of all genes identified and background list for GO enrichment analysis
#MD.ALL <- rbind(MD.Time, MD.D10, MD.D135, MD.D145)
MD.ALL.Genes <- unique(MD.All$gene)
MD.ALL.Genes <- sort(MD.ALL.Genes)
write.csv(MD.ALL.Genes, file = "Output/1.2_alignment/MD.ALL.Genes_For_GO_Background.csv")
```

# Testing for Differentially Methylated Genes
```{r}
# Comparison of samples between treatments
# Binomial GLM to test for differentially methylated genes
sub_meth_table  <- MD.All[,-15]
sub_meth_table$group <- paste0(sub_meth_table$Sample.ID, sub_meth_table$gene)
#filter for genes with >3 methylated positions
min.filt <- count(sub_meth_table, vars = c( group))
newdata <- min.filt[ which(min.filt$n > 2), ]
sub_meth_table <- sub_meth_table[sub_meth_table$group %in% newdata$vars,]
sub_meth_table <- sub_meth_table %>%
  mutate(sym = recode(trt1, b = "D", c = "C"),
         sym_trt = paste(sym, trt2, sep = ""))

# create data frame to stored results
results <- data.frame()
gs <- unique(sub_meth_table$gene)
#first subset the unique dataframes and second run the GLMs
for(i in 1:length(gs)){
  
  #subset the dataframe gene by gene
  sub_meth_table1 <- subset(sub_meth_table, gene ==gs[i])
  
  # fit glm position model
  fit <- glm(matrix(c(meth, unmeth), ncol=2) ~ sym*trt2, 
             data=sub_meth_table1, family=binomial)
  a <- anova(fit, test="Chisq")
  
  # capture summary stats to data frame
  df <- data.frame(gene = sub_meth_table1[1,7],
                   pval.symb = a$`Pr(>Chi)`[2],
                   pval.trt2 = a$`Pr(>Chi)`[3],
                   pval.symb_trt2 = a$`Pr(>Chi)`[4],
                   stringsAsFactors = F)
  
  # bind rows of temporary data frame to the results data frame
  results <- rbind(results, df)
  
}
# An error will be generated here for contrasts. 
#This potential for contrasts (interactions) is included in the case one wants to examine the role of position of CpG within a gene
#Error in `contrasts<-`(`*tmp*`, value = contr.funs[1 + isOF[nn]]) : contrasts can be applied only to factors with 2 or more levels
#Continuing the analysis from results line will generate the results in the absence of the contrast (interaction).
results[is.na(results)] <- 0
results$adj.pval.sym <- p.adjust(results$pval.symb, method='BH')
results$adj.pval.trt2 <- p.adjust(results$pval.trt2, method='BH')
results$adj.pval.sym_trt2 <- p.adjust(results$pval.symb_trt2, method='BH')

sig_genes <-results
sym_sig_genes <- sig_genes[, c(1,5)]
trt_sig_genes <- sig_genes[, c(1,6)]
inter_sig_genes <- sig_genes[, c(1,7)]
  
sym_sig_genes <- sym_sig_genes[order(sym_sig_genes$adj.pval.sym),]
sym_sig_genes <- sym_sig_genes[which(sym_sig_genes$adj.pval.sym <0.05), ]

trt_sig_genes <- trt_sig_genes[order(trt_sig_genes$adj.pval.trt2),]
trt_sig_genes <- trt_sig_genes[which(trt_sig_genes$adj.pval.trt2 <0.05), ]

inter_sig_genes <- inter_sig_genes[order(inter_sig_genes$adj.pval.sym_trt2),]
inter_sig_genes <- inter_sig_genes[which(inter_sig_genes$adj.pval.sym_trt2 <0.05), ]

sub_meth_table$per.meth <- (sub_meth_table$meth/(sub_meth_table$meth+sub_meth_table$unmeth))*100 #calculate percent methylation

# Annotation of DMG between symbionts
sym_DMG_annot <- merge(sym_sig_genes , Annot, by="gene", all.x=TRUE)
sym_DMG_annot <- sym_DMG_annot[order(sym_DMG_annot$adj.pval.sym),]
sym_DMG_annot <- sym_DMG_annot[!duplicated(sym_DMG_annot$gene),]
write.table(sym_DMG_annot, 'Output/1.2_alignment/Table_S1_Sym_DMG_annot.tsv', sep='\t', row.names=FALSE)

# Annotation of DMG between heat treatments
trt_DMG_annot <- merge(trt_sig_genes , Annot, by="gene", all.x=TRUE)
trt_DMG_annot <- trt_DMG_annot[order(trt_DMG_annot$adj.pval.trt2),]
trt_DMG_annot <- trt_DMG_annot[!duplicated(trt_DMG_annot$gene),]
write.table(sym_DMG_annot, 'Output/1.2_alignment/Table_S1_heat_DMG_annot.tsv', sep='\t', row.names=FALSE)

# Annotation of DMG between interaction
inter_DMG_annot <- merge(inter_sig_genes , Annot, by="gene", all.x=TRUE)
inter_DMG_annot <- inter_DMG_annot[order(inter_DMG_annot$adj.pval.sym_trt2),]
inter_DMG_annot <- inter_DMG_annot[!duplicated(inter_DMG_annot$gene),]
write.table(sym_DMG_annot, 'Output/1.2_alignment/Table_S1_SymXtrt2_DMG_annot.tsv', sep='\t', row.names=FALSE)

```


```{r}
##### Heatmap of DMGs  #####
meth_table.means <- aggregate(per.meth ~ treatment*gene, data=MD.All, FUN=mean)
meth_table.means_s <- meth_table.means[meth_table.means$gene %in% sym_sig_genes$gene,]
meth_table.means_t <- meth_table.means[meth_table.means$gene %in% trt_sig_genes$gene,]
meth_table.means_i <- meth_table.means[meth_table.means$gene %in% inter_sig_genes$gene,]

meth_table.means <- rbind(meth_table.means_t, meth_table.means_s)
meth_table.means <- rbind(meth_table.means, meth_table.means_i)
meth_table.means$tempID <- paste(meth_table.means$treatment, meth_table.means$gene, meth_table.means$per.meth, sep = ":")
meth_table.means <- meth_table.means[match(unique(meth_table.means$tempID), meth_table.means$tempID),]
meth_table.means <- meth_table.means[,-4]

sig_meth_mean <- meth_table.means %>% pivot_wider(names_from = c(treatment), values_from = per.meth)
sig_meth_mean.mat <- as.matrix(sig_meth_mean[,-1])
colnames(sig_meth_mean.mat) <- c("SymbC_Control", "SymbC_Heated", "SymbD_Control", "SymbD_Heated")
pdf("Output/1.2_alignment/DMG_HEATMAP.pdf")
AllHM <- heatmap.2(sig_meth_mean.mat, margins = c(10,10), cexRow = 1.2, cexCol = 1, Colv=FALSE, dendrogram="row", labRow=FALSE, col = rev(RColorBrewer::brewer.pal(name = "RdYlBu", n = 10)), na.color = "black", density.info = "none", trace = "none", scale = "row", sepwidth=c(0.025,0.025), sepcolor="darkgray", keysize = 1)
dev.off()
```


## GO enrichment
```{r}
all_GO_slim <- read.csv("Genome/Mcav_annotation_gene_slims.tab", sep = "\t")

##### GO enrichment of DMGs Sym #####
ALL<-as.data.frame(MD.ALL.Genes) #set the all gene list
colnames(ALL) <-c("gene")
DMG <- as.character(sym_sig_genes$gene) #set the enrichment test list
LENGTH <-cbind(Genes$gene,Genes$stop-Genes$start) #reads in a list of gene lengths
colnames(LENGTH) <-c("gene", "length") #name columns
LENGTH <- merge(LENGTH, ALL, by="gene") #merge gene names and length
GOAL <- merge(ALL, Gene.GO.IDs.slims,by="gene", all=T) #merge GO, ID
GOAL$GO.Slim.Term[is.na(GOAL$GO.Slim.Term)] <- "unknown" #list NA as unknown
splitted <- strsplit(as.character(GOAL$GO.IDs), "; ") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GOAL$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
GO.slimmed.terms <-GOAL[,c(1,4)]
#change contig lists to vectors
ALL.vector <-c(t(ALL))
DMG.vector <-c(t(DMG))
ID.vector <- LENGTH$gene
LENGTH.vector <-LENGTH$length
LENGTH.vector <- as.numeric(LENGTH.vector)
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=FALSE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
slim.cats <- Gene.GO.IDs.slims[,4:5]
slim.cats <- slim.cats[!duplicated(slim.cats$GO.Slim.Term), ]
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GOslim.Sym <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.Sym , file = "Output/1.2_alignment/GOslim_Sym.csv")
sig.GOslim.Sym <- GOslim.Sym %>%
  filter(over_represented_pvalue <0.08 | under_represented_pvalue<0.08)

##### GO enrichment of DMGs trt2 #####
ALL<-as.data.frame(MD.ALL.Genes) #set the all gene list
colnames(ALL) <-c("gene")
DMG <- as.character(trt_sig_genes$gene) #set the enrichment test list
LENGTH <-cbind(Genes$gene,Genes$stop-Genes$start) #reads in a list of gene lengths
colnames(LENGTH) <-c("gene", "length") #name columns
LENGTH <- merge(LENGTH, ALL, by="gene") #merge gene names and length
GOAL <- merge(ALL, Gene.GO.IDs.slims,by="gene", all=T) #merge GO, ID
GOAL$GO.Slim.Term[is.na(GOAL$GO.Slim.Term)] <- "unknown" #list NA as unknown
splitted <- strsplit(as.character(GOAL$GO.IDs), "; ") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GOAL$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
GO.slimmed.terms <-GOAL[,c(1,4)]
#change contig lists to vectors
ALL.vector <-c(t(ALL))
DMG.vector <-c(t(DMG))
ID.vector <- LENGTH$gene
LENGTH.vector <-LENGTH$length
LENGTH.vector <- as.numeric(LENGTH.vector)
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=FALSE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
slim.cats <- Gene.GO.IDs.slims[,4:5]
slim.cats <- slim.cats[!duplicated(slim.cats$GO.Slim.Term), ]
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GOslim.trt <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.trt , file = "Output/1.2_alignment/GOslim_trt.csv")
sig.GOslim.trt <- GOslim.trt %>%
  filter(over_represented_pvalue <0.08 | under_represented_pvalue<0.08)

##### GO enrichment of DMGs Sym:trt #####
ALL<-as.data.frame(MD.ALL.Genes) #set the all gene list
colnames(ALL) <-c("gene")
DMG <- as.character(inter_sig_genes$gene) #set the enrichment test list
LENGTH <-cbind(Genes$gene,Genes$stop-Genes$start) #reads in a list of gene lengths
colnames(LENGTH) <-c("gene", "length") #name columns
LENGTH <- merge(LENGTH, ALL, by="gene") #merge gene names and length
GOAL <- merge(ALL, Gene.GO.IDs.slims,by="gene", all=T) #merge GO, ID
GOAL$GO.Slim.Term[is.na(GOAL$GO.Slim.Term)] <- "unknown" #list NA as unknown
splitted <- strsplit(as.character(GOAL$GO.IDs), "; ") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GOAL$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
GO.slimmed.terms <-GOAL[,c(1,4)]
#change contig lists to vectors
ALL.vector <-c(t(ALL))
DMG.vector <-c(t(DMG))
ID.vector <- LENGTH$gene
LENGTH.vector <-LENGTH$length
LENGTH.vector <- as.numeric(LENGTH.vector)
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=FALSE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
slim.cats <- Gene.GO.IDs.slims[,4:5]
slim.cats <- slim.cats[!duplicated(slim.cats$GO.Slim.Term), ]
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GOslim.Inter <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.Inter , file = "Output/1.2_alignment/GOslim_inter.csv")
sig.GOslim.Inter <- GOslim.Inter %>%
  filter(over_represented_pvalue <0.08 | under_represented_pvalue<0.08)

##### GO enrichment of DMGs ALL #####
ALL<-as.data.frame(MD.ALL.Genes) #set the all gene list
colnames(ALL) <-c("gene")
DMG <- as.character(sig_meth_mean$gene) #set the enrichment test list
LENGTH <-cbind(Genes$gene,Genes$stop-Genes$start) #reads in a list of gene lengths
colnames(LENGTH) <-c("gene", "length") #name columns
LENGTH <- merge(LENGTH, ALL, by="gene") #merge gene names and length
GOAL <- merge(ALL, Gene.GO.IDs.slims,by="gene", all=T) #merge GO, ID
GOAL$GO.Slim.Term[is.na(GOAL$GO.Slim.Term)] <- "unknown" #list NA as unknown
splitted <- strsplit(as.character(GOAL$GO.IDs), "; ") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GOAL$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
GO.slimmed.terms <-GOAL[,c(1,5)]
#change contig lists to vectors
ALL.vector <-c(t(ALL))
DMG.vector <-c(t(DMG))
ID.vector <- LENGTH$gene
LENGTH.vector <-LENGTH$length
LENGTH.vector <- as.numeric(LENGTH.vector)
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=FALSE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
slim.cats <- Gene.GO.IDs.slims[,4:5]
slim.cats <- slim.cats[!duplicated(slim.cats$GO.Slim.Term), ]
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GOslim.All <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.All , file = "Output/1.2_alignment/GOslim_All.csv")
sig.GOslim.All <- GOslim.All %>%
  filter(over_represented_pvalue <0.08 | under_represented_pvalue<0.08)
```

Genes in Over and Under-represented GO Slim terms
```{r}
All.Slims <-  unique(Gene.GO.IDs.slims$GO.Slim.Term)
colnames(GO.slimmed.terms) <- c("gene", "GO.Slim.Term") 
nrow(sig.GOslim.All)
slim.genes.1 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[1,1]) #integral to plasma membrane 
slim.genes.2 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[2,1]) #cell division
slim.genes.3 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[3,1]) #extracellular region
slim.genes.4 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[4,1]) #tRNA processing
slim.genes.5 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[5,1]) #ubiquitin-specific protease activity
slim.genes.6 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[6,1]) #protein tyrosine kinase activity
slim.genes.7 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[7,1]) #motor activity
slim.genes.8 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[8,1]) #myosin complex
slim.genes.9 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[9,1]) #DNA integration
slim.genes.10 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[10,1]) #aspartic-type endopeptidase activity
slim.genes.11 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[11,1]) #protein binding
slim.genes.12 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[12,1]) #microtubule motor activity
slim.genes.13 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[13,1]) #ubiquitin-protein ligase activity 
slim.genes.14 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[14,1]) #RNA-directed DNA polymerase activity
slim.genes.15 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[15,1]) #zinc ion binding
slim.genes.16 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[16,1]) #identical protein binding
slim.genes.17 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[17,1]) #ion transport
slim.genes.18 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[18,1]) #negative regulation of apoptosis
slim.genes.19 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[19,1]) #membrane
slim.genes.20 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[20,1]) #nucleoplasm
slim.genes.21 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[21,1]) #nucleolus
slim.genes.22 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[22,1]) #mitochondrial inner membrane
slim.genes.23 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[23,1]) #integral to membrane
slim.genes.24 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[24,1]) #endoplasmic reticulum membrane
slim.genes.25 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[25,1]) #cytoplasmic mRNA processing body
slim.genes.26 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[26,1]) #mRNA processing
slim.genes.27 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[27,1]) #RNA splicing
slim.genes.28 <- GO.slimmed.terms %>% filter(GO.Slim.Term == sig.GOslim.All[28,1]) #translation



```

Plots of Methylation of Genes in Over and Under-represented GO Slim terms

```{r}
##### DMGs in all comparisons #####
ints_sym.trt<- merge(sym_sig_genes, trt_sig_genes, by="gene")
ints_sym.trt.inter <- merge(ints_sym.trt, inter_sig_genes, by="gene")
ints_all <- ints_sym.trt.inter
ints_all.annot <- merge(ints_all, Annot, by="gene", all.x=TRUE)
ints_all.annot <- ints_all.annot[!duplicated(ints_all.annot$gene),]
write.table(ints_all.annot, 'Output/1.2_alignment/Table_S5_DMGs_consistent_INT.tsv', sep='\t', row.names=FALSE)
```

```{r}
# GO Enrichment Analysis of DMG present in all comparisons
#### GO enrichment of DMGs #####
DMG <- as.character(ints_all$gene) #set the enrichment test list
#change contig lists to vectors
DMG.vector <-c(t(DMG))
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GOslim.INT_ALL <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.INT_ALL , file = "Output/1.2_alignment/GOslim_INT_ALL.csv")
sig.GOslim.INT.All <- GOslim.INT_ALL %>%
  filter(over_represented_pvalue <0.05 | under_represented_pvalue<0.05)


##### GO-BP enrichment of DMGs ALL #####
ALL<-as.data.frame(MD.ALL.Genes) #set the all gene list
colnames(ALL) <-c("gene")
DMG <- as.character(sig_meth_mean$gene) #set the enrichment test list
LENGTH <-cbind(Genes$gene,Genes$stop-Genes$start) #reads in a list of gene lengths
colnames(LENGTH) <-c("gene", "length") #name columns
LENGTH <- merge(LENGTH, ALL, by="gene") #merge gene names and length
GOAL <- merge(ALL, Gene.GO.IDs.slims[which(Gene.GO.IDs.slims$Cat == "P"),],by="gene", all=T) #merge GO, ID
GOAL$GO.Slim.Term[is.na(GOAL$GO.Slim.Term)] <- "unknown" #list NA as unknown
splitted <- strsplit(as.character(GOAL$GO.IDs), "; ") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(GOAL$gene, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row 
GO.slimmed.terms <-GOAL[,c(1,4)]
#change contig lists to vectors
ALL.vector <-c(t(ALL))
DMG.vector <-c(t(DMG))
ID.vector <- LENGTH$gene
LENGTH.vector <-LENGTH$length
LENGTH.vector <- as.numeric(LENGTH.vector)
gene.vector=as.integer(ALL.vector%in%DMG.vector) #Construct new vector with 1 for DEG and 0 for others
names(gene.vector)=ALL.vector #set names
DEG.pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) #weight vector by length of gene
#Find enriched GO terms, 
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO.slim.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.slimmed.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=FALSE)
colnames(GO.slim.wall)[1] <- "GO.Slim.Term"
slim.cats <- Gene.GO.IDs.slims[,4:5]
slim.cats <- slim.cats[!duplicated(slim.cats$GO.Slim.Term), ]
GO.slim.wall <- left_join(GO.slim.wall, slim.cats)
#How many enriched GO terms do we have
class(GO.slim.wall)
head(GO.slim.wall)
tail(GO.slim.wall)
nrow(GO.slim.wall)
GO_BP.slim.All <- GO.slim.wall[order(GO.slim.wall$over_represented_pvalue),]
write.csv(GOslim.All , file = "Output/1.2_alignment/GO_BP.slim_All.csv")
sig.GO_BP.slim.All <- GO_BP.slim.All %>%
  filter(over_represented_pvalue <0.08 | under_represented_pvalue<0.08)

```


